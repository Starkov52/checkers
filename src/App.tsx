import { useState } from "react";
import reactLogo from "./assets/react.svg";
import React from "react";
import viteLogo from "/vite.svg";
import "./css/main.css";
import Board from "./components/board";
import MainPage from "./components/mainPage";
import Rules from "./components/rules";
import ListServers from "./components/listServers";
import CreateServer from "./components/createServer";
import ServerBoard from "./components/serverBoard";
import { BrowserRouter as Router, Link, Route, Routes } from "react-router-dom";
import JoinBoard from "./components/joinBoard";
import JoinById from "./components/joinById";
function App() {
 const [count, setCount] = useState(0);

 return (
  <div className="AppClass">
   <Router>
    <Routes>
     <Route path="/" element={<MainPage></MainPage>}></Route>
     <Route path="/rules" element={<Rules></Rules>}></Route>
     <Route path="/localGame" element={<Board></Board>}></Route>
     <Route path="/listOnlineGame" element={<ListServers></ListServers>}></Route>
     <Route path="/listOnlineGame/createServer" element={<CreateServer></CreateServer>}></Route>
     <Route path="/listOnlineGame/createServer/game" element={<ServerBoard></ServerBoard>}></Route>
     <Route path="/listOnlineGame/joinServer/:id" element={<JoinBoard></JoinBoard>}></Route>
     <Route path="/listOnlineGame/joinById" element={<JoinById></JoinById>}></Route>
    </Routes>
   </Router>
  </div>
 );
}

export default App;

