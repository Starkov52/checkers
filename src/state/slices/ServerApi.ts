import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";

import { Board } from "../../components/board";
import { Server } from "./serverBoardSlice";
const serverApi = createApi({
 reducerPath: "api",
 baseQuery: fetchBaseQuery({ baseUrl: "https://telegrambotfishcombat-default-rtdb.firebaseio.com" }),
 endpoints: (builder) => ({
  sendServer: builder.mutation<void, Server>({
   query: (serverBoard: Server) => ({
    url: "/checkers.json",
    method: "POST",
    body: serverBoard,
    headers: {
     "Content-Type": "application/json",
    },
   }),
  }),
  getServer: builder.query<{}, string>({
   query: (path: string) => ({
    url: `/checkers.json`,
    method: "GET",
    headers: {
     "Content-Type": "application/json",
    },
   }),
  }),
  getDataById: builder.query<Server, string>({
   query: (path: string) => ({
    url: `/checkers/${path}.json`,
    method: "GET",
    headers: {
     "Content-Type": "application/json",
    },
   }),
  }),
 }),
});
export const { useSendServerMutation, useGetServerQuery, useGetDataByIdQuery } = serverApi;
export default serverApi;
