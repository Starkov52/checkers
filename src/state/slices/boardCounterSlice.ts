import { createSlice } from "@reduxjs/toolkit";
import store from "../store";
import { UseSelector } from "react-redux";
import { RootState } from "../store";
import { PayloadAction } from "@reduxjs/toolkit";
import { Board } from "../../components/board";
type counter = {
 white: number;
 black: number;
};
const counterP: counter = {
 white: 0,
 black: 0,
};

const counterSlice = createSlice({
 name: "counter",
 initialState: counterP,
 reducers: {
  calculatePoints: (state: counter, action: PayloadAction<Board>) => {
   const board = action.payload;
   let sumBlack: number = 0;
   let sumWhite: number = 0;
   for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[i].length; j++) {
     if (board[i][j].check.colorChess === "black") {
      if (board[i][j].check.chessStatus !== false) {
       sumBlack += 1;
      }
     } else if (board[i][j].check.colorChess === "white") {
      if (board[i][j].check.chessStatus !== false) {
       sumWhite += 1;
      }
     }
    }
   }
   state.black = sumBlack;
   state.white = sumWhite;
  },
 },
});
export const counter = counterSlice.reducer;
export const { calculatePoints } = counterSlice.actions;
