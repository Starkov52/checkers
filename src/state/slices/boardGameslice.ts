import { createSlice } from "@reduxjs/toolkit";
import { Check, Board, Row, Cube } from "../../components/board";
import { RootState } from "../store";
import { PayloadAction } from "@reduxjs/toolkit";
enum chess {
 default = "default",
 king = "king",
}
export const stateBoard: Board = [
 [
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
 ],
 [
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
 ],
 [
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "black" },
   isSelected: false,
  },
 ],
 [
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
 ],
 [
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  { color: "black", check: { chessStatus: false, colorChess: false }, isSelected: false },
 ],
 [
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
 ],
 [
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
 ],
 [
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
  {
   color: "black",
   check: { chessStatus: chess.default, colorChess: "white" },
   isSelected: false,
  },
  { color: "brown", check: { chessStatus: false, colorChess: false }, isSelected: false },
 ],
];
export const gameBoardSlice = createSlice({
 name: "board",
 initialState: stateBoard,
 reducers: {
  updateBoardLocal: (state: Board, action: PayloadAction<Board>) => {
   console.log("НОВОЕ СОСТОЯНИЕ REDUX", action.payload[4][3]);
   return [...action.payload];
  },
  resetBoard: (state: Board, action: PayloadAction<Board>) => {
   return [...action.payload];
  },
 },
});

export const board = gameBoardSlice.reducer;
export const { updateBoardLocal, resetBoard } = gameBoardSlice.actions;
