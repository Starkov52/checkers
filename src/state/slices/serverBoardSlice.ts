import { createSlice } from "@reduxjs/toolkit";
import { stateBoard } from "./boardGameslice";
import { Board } from "../../components/board";
import { PayloadAction } from "@reduxjs/toolkit";
import { createAsyncThunk } from "@reduxjs/toolkit";

interface serverState {
 loading: boolean;
 server: null;

 error: null | boolean;
 success: false;
}
export type Server = {
 board: Board;
 hostName: string;
 serverName: string;
 step: boolean;
 guest: string | null;
 id: string;
 serverState: serverState;
};
const serverData: Server = {
 board: stateBoard,
 hostName: "–ê–Ω–æ–Ω–∏–º__ " + Math.floor(Math.random() * 10) + Math.floor(Math.random() * 10) + Math.floor(Math.random() * 10) + Math.floor(Math.random() * 10),
 serverName: "–°–µ—Ä–≤–µ—Ä",
 step: true,
 guest: "–Ω–æ—É–Ω–µ–π–º",
 id: "none",
 serverState: {
  loading: false,
  server: null,

  error: null,
  success: false,
 },
};

export const postF = createAsyncThunk<void, { path: string; body: Server }, { rejectValue: string }>("joinTheGame", async ({ path, body }: { path: string; body: Server }) => {
 try {
  const fetchQuery = await fetch(`https://telegrambotfishcombat-default-rtdb.firebaseio.com/checkers/${path}.json`, {
   method: "PUT",
   body: JSON.stringify(body),
   headers: {
    "Content-Type": "application/json",
   },
  }).then((response) => {
   if (response.ok) {
    return response.json();
   } else {
    throw new Error("error to fetch data session game");
   }
  });
 } catch (error: any) {
  console.error(error.message);
 }
});
export const getF = createAsyncThunk<Server, { path: string }>("update", async ({ path }: { path: string }) => {
 try {
  const response = await fetch(`https://telegrambotfishcombat-default-rtdb.firebaseio.com/checkers/${path}.json`, {
   headers: {
    "Content-Type": "application/json",
   },
   method: "GET",
  });
  if (response.ok) {
   const rresponse: Server = await response.json();
   return rresponse;
  } else {
   throw new Error("Error to get data");
  }
 } catch (error: any) {
  throw new Error(error);
 }
});
export const getFN = createAsyncThunk<Server, { path: string }>("update", async ({ path }: { path: string }) => {
 try {
  const response = await fetch(`https://telegrambotfishcombat-default-rtdb.firebaseio.com/checkers/${path}.json`, {
   headers: {
    "Content-Type": "application/json",
   },
   method: "GET",
  });
  if (response.ok) {
   const rresponse: Server = await response.json();
   return rresponse;
  } else {
   throw new Error("Error to get data");
  }
 } catch (error: any) {
  throw new Error(error);
 }
});

const serverSlice = createSlice({
 name: "server",
 initialState: serverData,
 reducers: {
  updateServerName: (state, action: PayloadAction<string>) => {
   console.log("–†–µ–¥—é—Å–µ—Ä updateServerName —Å—Ä–∞–±–æ—Ç–∞–ª:", action.payload);
   state.serverName = action.payload;
  },
  updateBoard: (state, board: PayloadAction<Board>) => {
   state.board = board.payload;
   state.step = !state.step;
  },
 },
 extraReducers: (builder) => {
  builder
   .addCase(postF.pending, (state) => {
    state.serverState.loading = true;
   })
   .addCase(postF.fulfilled, (state) => {
    state.serverState.loading = false;
    state.serverState.error = false;
   })
   .addCase(postF.rejected, (state) => {
    state.serverState.error = true;
   })
   .addCase(getF.pending, (state) => {
    state.serverState.loading = true;
   })
   .addCase(getF.fulfilled, (state, action: PayloadAction<Server>) => {
    console.log("üî• –î–∞–Ω–Ω—ã–µ –ø–æ–ª—É—á–µ–Ω—ã –∏–∑ getF:", action.payload); // –õ–æ–≥ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞ –∑–∞–ø—Ä–æ—Å–∞

    return {
     ...state,
     board: action.payload.board,
     id: action.payload.id,
     step: action.payload.step,
     guest: action.payload.guest,
     hostName: action.payload.hostName,
     serverState: { ...state.serverState, loading: false, error: false },
    };
   })
   .addCase(getF.rejected, (state) => {
    state.serverState.error = true;
   });
 },
});
export const { updateServerName, updateBoard } = serverSlice.actions;
export const serverBoard = serverSlice.reducer;
