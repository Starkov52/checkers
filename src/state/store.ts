import { gameBoardSlice } from "./slices/boardGameslice";
import { configureStore } from "@reduxjs/toolkit";
import { combineReducers } from "@reduxjs/toolkit";
import { board } from "./slices/boardGameslice";
import { counter } from "./slices/boardCounterSlice";
import { serverBoard } from "./slices/serverBoardSlice";
import serverApi from "./slices/ServerApi";

const reducers = combineReducers({
 gameBoard: board,
 gameCounter: counter,
 gameServerBoard: serverBoard,
 [serverApi.reducerPath]: serverApi.reducer,
});

const store = configureStore({
 reducer: reducers,
 middleware: (getDefaultMiddleware) => {
  return getDefaultMiddleware().concat(serverApi.middleware);
 },
});
export type RootState = ReturnType<typeof store.getState>;
export type RootDispatxh = typeof store.dispatch;
export default store;
