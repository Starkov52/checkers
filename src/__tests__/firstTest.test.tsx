import Board from "../components/board";
import React from "react";
import App from "../App";
import { render, screen, waitFor, within } from "@testing-library/react";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import { Provider, useDispatch } from "react-redux";
import { resetBoard } from "../state/slices/boardGameslice";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import dosk from "../2c1713eb-cfa5-48b2-88a3-603be26f86a4.png";
import Rules from "../components/rules";
import ListServers from "../components/listServers";
import { stateBoard } from "../state/slices/boardGameslice";
import CreateServer from "../components/createServer";
import store, { RootDispatxh } from "../state/store";
import ServerBoard from "../components/serverBoard";
import JoinBoard from "../components/joinBoard";
import JoinById from "../components/joinById";
import userEvent from "@testing-library/user-event";
import MainPage from "../components/mainPage";
import { hostname } from "os";
enum chess {
 default = "default",
 king = "king",
}

describe("first", () => {
 let currentState = store.getState();

 beforeEach(() => {
  const bboard = [...stateBoard];
  store.dispatch(resetBoard(bboard));
  render(
   <Provider store={store}>
    <BrowserRouter>
     <Board></Board>
    </BrowserRouter>
   </Provider>,
  );
 });

 test("UP-RIGHT", async () => {
  const board = screen.getByTestId("board");
  expect(board).toBeInTheDocument();

  const firstChecker = screen.getByTestId("cell-5-0");
  expect(firstChecker).toBeInTheDocument();
  userEvent.click(firstChecker);

  const fieldToSend = screen.getByTestId("cell-4-1");

  expect(fieldToSend).toBeInTheDocument();
  userEvent.click(fieldToSend);
  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;

   expect(updatedBoard[4][1].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });
 });
 test("UP-LEFT", async () => {
  const board = screen.getByTestId("board");
  expect(board).toBeInTheDocument();

  const firstChecker = screen.getByTestId("cell-5-2");
  expect(firstChecker).toBeInTheDocument();
  userEvent.click(firstChecker);

  const fieldToSend = screen.getByTestId("cell-4-1");

  expect(fieldToSend).toBeInTheDocument();
  userEvent.click(fieldToSend);
  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;

   expect(updatedBoard[4][1].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });
 });
 test("UP-FIGHT-LEFT", async () => {
  const board = screen.getByTestId("board");
  expect(board).toBeInTheDocument();

  const firstChecker = screen.getByTestId("cell-5-2");
  expect(firstChecker).toBeInTheDocument();
  userEvent.click(firstChecker);

  const fieldToSend = screen.getByTestId("cell-4-3");
  expect(fieldToSend).toBeInTheDocument();
  userEvent.click(fieldToSend);
  const filedCheckerWhite = screen.getByTestId("cell-2-1");
  userEvent.click(filedCheckerWhite);
  const fieldToSendWhite = screen.getByTestId("cell-3-2");
  userEvent.click(fieldToSendWhite);

  userEvent.click(fieldToSend);
  const fieldToSendBlack = screen.getByTestId("cell-2-1");
  userEvent.click(fieldToSendBlack);
  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;
   expect(updatedBoard[2][1].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });
 });

 test("DOWN-FIGHT-RIGHT", async () => {
  const board = screen.getByTestId("board");
  expect(board).toBeInTheDocument();

  const firstChecker = screen.getByTestId("cell-5-2");
  expect(firstChecker).toBeInTheDocument();
  await userEvent.click(firstChecker);

  const fieldToSend = screen.getByTestId("cell-4-3");
  expect(fieldToSend).toBeInTheDocument();
  await userEvent.click(fieldToSend);
  const filedCheckerWhite = screen.getByTestId("cell-2-1");
  expect(filedCheckerWhite).toBeInTheDocument();
  await userEvent.click(filedCheckerWhite);
  const fieldToSendWhite = screen.getByTestId("cell-3-2");
  expect(fieldToSendWhite).toBeInTheDocument();
  await userEvent.click(fieldToSendWhite);

  await userEvent.click(fieldToSend);
  const fieldToSendBlack = screen.getByTestId("cell-2-1");
  expect(fieldToSendBlack).toBeInTheDocument();
  await userEvent.click(fieldToSendBlack);
  //
  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;
   expect(updatedBoard[2][1].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });

  const fieldToSendBlackToDie = screen.getByTestId("cell-2-3");
  expect(fieldToSendBlackToDie).toBeInTheDocument();
  await userEvent.click(fieldToSendBlackToDie);
  await userEvent.click(fieldToSendWhite);
  await userEvent.click(fieldToSendBlack);
  await userEvent.click(fieldToSend);
  await new Promise((resolve) => setTimeout(resolve, 100));
  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;
   console.log(updatedBoard[4][3].check);
   screen.debug();
   expect(updatedBoard[4][3].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });
 });
 test("DOWN-FIGHT-LEFT", async () => {
  const board = screen.getByTestId("board");
  expect(board).toBeInTheDocument();

  const firstChecker = screen.getByTestId("cell-5-2");
  expect(firstChecker).toBeInTheDocument();
  await userEvent.click(firstChecker);

  const fieldToSend = screen.getByTestId("cell-4-3");
  expect(fieldToSend).toBeInTheDocument();
  await userEvent.click(fieldToSend);
  const filedCheckerWhite = screen.getByTestId("cell-2-1");
  expect(filedCheckerWhite).toBeInTheDocument();
  await userEvent.click(filedCheckerWhite);
  const fieldToSendWhite = screen.getByTestId("cell-3-2");
  expect(fieldToSendWhite).toBeInTheDocument();
  await userEvent.click(fieldToSendWhite);

  await userEvent.click(fieldToSend);
  const fieldToSendBlack = screen.getByTestId("cell-2-1");
  expect(fieldToSendBlack).toBeInTheDocument();
  await userEvent.click(fieldToSendBlack);
  //
  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;
   expect(updatedBoard[2][1].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });

  const fieldToSendBlackToDie = screen.getByTestId("cell-2-3");
  expect(fieldToSendBlackToDie).toBeInTheDocument();
  await userEvent.click(fieldToSendBlackToDie);
  await userEvent.click(fieldToSendWhite);
  await userEvent.click(fieldToSendBlack);
  await userEvent.click(fieldToSend);

  await waitFor(() => {
   const updatedBoard = store.getState().gameBoard;
   console.log(updatedBoard[4][3].check);
   screen.debug();
   expect(updatedBoard[4][3].check).toEqual({
    chessStatus: chess.default,
    colorChess: "white",
   });
  });
  //

  const fieltToSendBlack = screen.getByTestId("cell-1-2");
  await userEvent.click(fieltToSendBlack);
  const fieltToSend = screen.getByTestId("cell-2-1");
  await userEvent.click(fieltToSend);
  await userEvent.click(fieldToSend);
  const cell = screen.getByTestId("cell-3-4");
  await userEvent.click(cell);
  await userEvent.click(fieltToSend);
  await userEvent.click(fieldToSendWhite);
  await userEvent.click(cell);
  await userEvent.click(fieldToSendBlackToDie);
  const xuyna = screen.getByTestId("cell-1-0");
  await userEvent.click(xuyna);
  await userEvent.click(fieltToSend);
  await userEvent.click(fieldToSendBlackToDie);
  const killBlack = screen.getByTestId("cell-4-1");
  await userEvent.click(killBlack);

  const updatedBoard = store.getState().gameBoard;
  expect(updatedBoard[4][1].check).toEqual({
   chessStatus: chess.default,
   colorChess: "white",
  });
 });
});
describe("allRoutingTests", () => {
 beforeEach(() => {
  const { asFragment } = render(
   <Provider store={store}>
    <MemoryRouter initialEntries={["/"]}>
     <Routes>
      <Route path="/" element={<MainPage></MainPage>}></Route>
      <Route path="/rules" element={<Rules></Rules>}></Route>
      <Route path="/localGame" element={<Board></Board>}></Route>
      <Route path="/listOnlineGame" element={<ListServers></ListServers>}></Route>
      <Route path="/listOnlineGame/createServer" element={<CreateServer></CreateServer>}></Route>
      <Route path="/listOnlineGame/createServer/game" element={<ServerBoard></ServerBoard>}></Route>
      <Route path="/listOnlineGame/joinServer/:id" element={<JoinBoard></JoinBoard>}></Route>
      <Route path="/listOnlineGame/joinById" element={<JoinById></JoinById>}></Route>
     </Routes>
    </MemoryRouter>
   </Provider>,
  );
  expect(asFragment()).toMatchSnapshot();
 });
 test("routing", async () => {
  const mainPage = await screen.getByTestId("mainPage");
  expect(mainPage).toBeInTheDocument();
  const link = await screen.getByTestId("localGame");
  await userEvent.click(link);
  const board = await screen.getByTestId("board");
  expect(board).toBeInTheDocument();
  const closeBtn = await screen.getByTestId("close");
  await userEvent.click(closeBtn);
  await expect(await screen.getByTestId("mainPage")).toBeInTheDocument();
  const rules = await screen.getByTestId("rulesBtn");
  await expect(rules).toBeInTheDocument();
  await userEvent.click(rules);

  const rulesPage = await screen.getByTestId("rules");
  expect(rulesPage).toBeInTheDocument();
  const closeRules = await screen.getByTestId("rulesClose");
  await userEvent.click(closeRules);
  await expect(screen.getByTestId("mainPage")).toBeInTheDocument();
  const serversBtn = await screen.getByTestId("serverBtn");
  await userEvent.click(serversBtn);
  await expect(screen.getByTestId("server")).toBeInTheDocument();
  const serverClose = await screen.getByTestId("serverClose");
  await userEvent.click(serverClose);
  await expect(screen.getByTestId("mainPage")).toBeInTheDocument();
 });
 test("testCreateServer", async () => {
  const serversBtn = await screen.getByTestId("serverBtn");
  await userEvent.click(serversBtn);
  const addServerBtn = await screen.getByTestId("addServer");
  await userEvent.click(addServerBtn);
  const inputServer = await screen.getByTestId("inputServer");

  expect(inputServer).toBeInTheDocument();
  await userEvent.type(inputServer, "123");

  expect(inputServer).toHaveValue("123");
 });
});
describe("fetchTests", () => {
 beforeAll(() => {
  global.fetch = jest.fn(() =>
   Promise.resolve({
    ok: true,
    json: () =>
     Promise.resolve([
      {
       board: stateBoard,
       hostName: "Новый Орлеан",
       serverName: "123123",
       step: true,
       guest: null,
       id: "dqwqwdqwd",
       serverState: { error: false, loading: false, success: false },
      },
      {
       board: stateBoard,
       hostName: "Новы12eй Орлеан",
       serverName: "1232112123",
       step: true,
       guest: null,
       id: "dqwqwdq2112wd",
       serverState: { error: false, loading: false, success: false },
      },
     ]),
   } as Response),
  );
 });

 beforeEach(() => {
  const { asFragment } = render(
   <Provider store={store}>
    <MemoryRouter initialEntries={["/"]}>
     <Routes>
      <Route path="/" element={<MainPage></MainPage>}></Route>
      <Route path="/rules" element={<Rules></Rules>}></Route>
      <Route path="/localGame" element={<Board></Board>}></Route>
      <Route path="/listOnlineGame" element={<ListServers></ListServers>}></Route>
      <Route path="/listOnlineGame/createServer" element={<CreateServer></CreateServer>}></Route>
      <Route path="/listOnlineGame/createServer/game" element={<ServerBoard></ServerBoard>}></Route>
      <Route path="/listOnlineGame/joinServer/:id" element={<JoinBoard></JoinBoard>}></Route>
      <Route path="/listOnlineGame/joinById" element={<JoinById></JoinById>}></Route>
     </Routes>
    </MemoryRouter>
   </Provider>,
  );
 });
 afterEach(() => {
  jest.restoreAllMocks();
 });
 test("one", async () => {
  // Ждем появления кнопки
  const mainPage = await screen.getByTestId("mainPage");
  expect(mainPage).toBeInTheDocument();
  const serversBtn = await screen.findByTestId("serverBtn");
  expect(serversBtn).toBeInTheDocument();

  // Клик и ожидание навигации
  await userEvent.click(serversBtn);

  // Ждем загрузки данных и появления элементов
  const serverElements = await screen.findAllByTestId("server");
  await expect(serverElements).toHaveLength(1);
  await expect(fetch).toHaveBeenCalledTimes(1);
 });
});
