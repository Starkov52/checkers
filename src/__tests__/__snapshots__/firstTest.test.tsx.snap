// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allRoutingTests routing 1`] = `
<DocumentFragment>
  <div
    class="mainPage"
    data-testid="mainPage"
  >
    <div
      class="mainPage__title"
    >
      <h1
        class="mainPage__titleT"
      >
        Checkers
      </h1>
      <h2
        class="mainPage__titleD"
      >
        deluxe edition
      </h2>
    </div>
    <div
      class="mainPage__buttons"
    >
      <a
        data-discover="true"
        href="/localGame"
      >
        <button
          class="mainPage__gameFriendBtn"
          data-testid="localGame"
        >
          Игра друг с друaaaa
        </button>
      </a>
      <a
        data-discover="true"
        href="/listOnlineGame"
      >
        <button
          class="mainPage__gameOnlineBtn"
          data-testid="serverBtn"
        >
          Сетевая игрa
        </button>
      </a>
      <a
        data-discover="true"
        href="/rules"
      >
        <button
          class="mainPage__gameRulesBtn"
          data-testid="rulesBtn"
        >
          Правила
        </button>
      </a>
    </div>
    <p
      class="mainPage__programmer"
    >
      By 
      <span
        style="color: black;"
      >
         Starkov 
      </span>
      03.
      <span
        style="font-family: monospace; color: black;"
      >
        2025
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`allRoutingTests testCreateServer 1`] = `
<DocumentFragment>
  <div
    class="mainPage"
    data-testid="mainPage"
  >
    <div
      class="mainPage__title"
    >
      <h1
        class="mainPage__titleT"
      >
        Checkers
      </h1>
      <h2
        class="mainPage__titleD"
      >
        deluxe edition
      </h2>
    </div>
    <div
      class="mainPage__buttons"
    >
      <a
        data-discover="true"
        href="/localGame"
      >
        <button
          class="mainPage__gameFriendBtn"
          data-testid="localGame"
        >
          Игра друг с друaaaa
        </button>
      </a>
      <a
        data-discover="true"
        href="/listOnlineGame"
      >
        <button
          class="mainPage__gameOnlineBtn"
          data-testid="serverBtn"
        >
          Сетевая игрa
        </button>
      </a>
      <a
        data-discover="true"
        href="/rules"
      >
        <button
          class="mainPage__gameRulesBtn"
          data-testid="rulesBtn"
        >
          Правила
        </button>
      </a>
    </div>
    <p
      class="mainPage__programmer"
    >
      By 
      <span
        style="color: black;"
      >
         Starkov 
      </span>
      03.
      <span
        style="font-family: monospace; color: black;"
      >
        2025
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`fetchTests one 1`] = `
<DocumentFragment>
  <div
    class="mainPage"
    data-testid="mainPage"
  >
    <div
      class="mainPage__title"
    >
      <h1
        class="mainPage__titleT"
      >
        Checkers
      </h1>
      <h2
        class="mainPage__titleD"
      >
        deluxe edition
      </h2>
    </div>
    <div
      class="mainPage__buttons"
    >
      <a
        data-discover="true"
        href="/localGame"
      >
        <button
          class="mainPage__gameFriendBtn"
          data-testid="localGame"
        >
          Игра друг с друaaaa
        </button>
      </a>
      <a
        data-discover="true"
        href="/listOnlineGame"
      >
        <button
          class="mainPage__gameOnlineBtn"
          data-testid="serverBtn"
        >
          Сетевая игрa
        </button>
      </a>
      <a
        data-discover="true"
        href="/rules"
      >
        <button
          class="mainPage__gameRulesBtn"
          data-testid="rulesBtn"
        >
          Правила
        </button>
      </a>
    </div>
    <p
      class="mainPage__programmer"
    >
      By 
      <span
        style="color: black;"
      >
         Starkov 
      </span>
      03.
      <span
        style="font-family: monospace; color: black;"
      >
        2025
      </span>
    </p>
  </div>
</DocumentFragment>
`;
